### ISSUE TITLE
[Deployment Issue] Deploy process takes forever recently with a bitbucket dependency

### ISSUE THREAD

**Issue by drmrbrewer** [caprover/caprover#2134]
I know this likely isn't anything to do with CapRover, though I did upgrade recently to that latest CapRover release and it's possible that this issue started happening around then.  But I can't be sure.

I have a dependency in the `package.json` file for a node.js app deployed with CapRover using a Dockerfile that itself calls `RUN npm install`.  The dependency is expressed in the following form in `package.json`:

    "dependencies": {
      "myrepo": "git+https://myusername:myapppassword@bitbucket.org/myusername/myrepo.git#v123"
    },
 
This has worked fine for years.  It still works, but I'm finding that recently that the deploy process hangs for a looooong time (around 10 minutes) on the `RUN npm install` step.  In the past it was maybe a few seconds... not really sure but never enough to notice particularly.

The bitbucket repo dependency is the only dependency, so it must relate to that?  When I view the server stats, it seems like nothing is really going on... CPU and network activity flatlining... so it seems like it is just waiting.

Any ideas why it's taking so long recently?

**Comment by githubsaturn** [comment]
The build process is handled fully by Docker, not CapRover. You can try building your docker image on server without CapRover and you'll likely see the same result. 

With regards to the cause, there are many factors that can cause that. To name a few:
- A certificate issue with your base image (FROM line) - try upgrading to a newer version.
- Any change in network configuration of your host
- Any DNS changes in your host's zone
- Any changes from Github to limit traffic from the band of a particular VPS provider, because someone else is attacking github from the same IP range. 
- and many more.


There are many similar questions online about the same issue:
- https://github.com/nodejs/docker-node/issues/1946
- https://www.reddit.com/r/docker/comments/11ylxf5/container_hangs_when_running_npm_i_ie_install_npm/
- https://stackoverflow.com/questions/56295037/docker-stuck-at-npm-install
- https://stackoverflow.com/questions/62556972/docker-image-build-getting-stuck-at-npm-install-for-reactjs
- https://stackoverflow.com/questions/63238207/npm-install-hangs-in-docker-container
- https://stackoverflow.com/questions/69575465/why-does-docker-gets-stuck-after-running-npm-install

**Comment by drmrbrewer** [comment]
Thanks for the insights, @githubsaturn .  Running with `npm install --verbose` provides some clues... firstly it was as follows:

    ... quite few npm lines above, then...
    npm http fetch GET 200 https://registry.npmjs.org/common 713ms (cache miss)
    npm http fetch GET 200 https://registry.npmjs.org/puppeteer-core/-/puppeteer-core-21.11.0.tgz 4060ms (cache miss)
    npm info run puppeteer@21.11.0 postinstall node_modules/puppeteer node install.mjs
    npm info run weacast-grib2json@1.0.3 postinstall node_modules/weacast-grib2json node set_permission
    npm info run weacast-grib2json@1.0.3 postinstall { code: 0, signal: null }
    ... hangs for AGES here... 

When I removed the `weacast-grib2json` dependency completely, it now seems to hang after:

    npm http fetch GET 200 https://registry.npmjs.org/common 732ms (cache miss)

In any case, it seems like it's some sort of rate limiting from `npm`, if this is even possible?  Maybe the IP address range of my cloud server has appeared on a blacklist.

**Comment by githubsaturn** [comment]
Definitely possible - I've seen this happening in the past. You can try using yarn instead of npm so that it uses `registry.yarnpkg.com`

**Comment by drmrbrewer** [comment]
Even with yarn I'm seeing massive delays.  But the verbose output tells me that it's spending most time (like 10 mins) on step `[4/4] Building fresh packages...`, which I assume to be a "local" stage that wouldn't be rate-limited by yarn.  The earlier fetching steps (e.g. `Performing "GET" request to "https://registry.yarnpkg.com/postgres-array/-/postgres-array-2.0.0.tgz"`) happen quickly.  So I'm really perplexed as to why it's taking 10 mins recently when it used to be super quick.  Ah well, thanks for your suggestions!

**Comment by drmrbrewer** [comment]
Eventually tracked it down to one of the dependencies (puppeteer) downloading some very large chromium-related binaries and other files.  Not sure why this wasn't an issue until recently, because I've remained on the same puppeteer version for over a year now, but hey... solution was to prevent this stuff from being downloaded with the following just before `npm install`:

    ENV PUPPETEER_SKIP_DOWNLOAD=true

**Comment by githubsaturn** [comment]
Ah interesting! Are you using npm i or npm ci?

The former ignores package-lock. You should always use npm ci

**Comment by drmrbrewer** [comment]
I'm using `npm i`... I don't use any `package-lock`.

**Comment by githubsaturn** [comment]
That's the problem! By using npm i, you're allowing future versions to be downloaded. Using nom ci makes sure you always have a specific version

**Comment by drmrbrewer** [comment]
But with `"puppeteer": "^21.11.0"` in package.json and with `21.11.0` being the last one before `22.0.0` I should always be getting `21.11.0`?

**Comment by githubsaturn** [comment]
npm has a complex dependency resolution logic. I am not an expert, but I know there is peer dependencies, etc... 
So there might be other dependencies that got updated and caused this....


Using `npm i` is definitely a recipe for problems in the future.

**Comment by drmrbrewer** [comment]
Ah yes OK I see.  I'll bear that in mind :)  Even if that dependency was fixed on `21.11.0` throughout the build, including peer dependencies, I did wonder whether maybe the downloading stage (which is where it was taking time) was just referring to a fixed link, but the structure at the far end of that link was changed recently (to add many more files / larger files).  

I might also switch to `yarn` based on your suggestion... I hadn't realised that (a) it's basically a drop-in replacement for `npm`, with the same packages and even the command-line options... at least the basic ones that I'm using; and (b) it's faster.
